### PR TITLE
Embedding gradient performance optimization on GPU

### PR DESCRIPTION
**Pull request by MoisesHer**
## Description ##
This PR includes a specific Embedding-backward operator for GPU.
Two new CUDA kernels have been implemented for improving the performance of the operator when using GPU.
According to our measurements on Volta GPUs, the previous version was taken 2.2ms,
whereas the new implementation takes 0.3ms, i.e. more than 7x speedup.
### Essentials ###
- [X ] Changes are complete (i.e. I finished coding on this PR)
- [X] All changes have test coverage:
- Unit tests already existed (tests/python/gpu/test_operator_gpu:test_embedding_with_type) and changes do not affect correctness.
- [X] Code is well-documented:
- For new C++ functions in header files, their functionalities and arguments are documented.
### Changes ###
- [x ] Embedding-backward operator for GPU, test: tests/python/gpu/test_operator_gpu:test_embedding_with_type

### CODE REVIEW

File: src/operator/tensor/indexing_op.cu [hunk 1]
```diff
@@ -545,6 +545,245 @@ void TakeOpForward<gpu>(const nnvm::NodeAttrs& attrs,
...
 }
 
+/*
+ * \brief returns integer log2(a) rounded up
+ */
+inline int ilog2(unsigned int a) {
+  int k = 1;
+  while (a >>= 1) k++;
+  return k;
+}
```
Comment (ptrendx) [inline src/operator/tensor/indexing_op.cu]: Please move this function to unnamed namespace.

**Comment by ptrendx** [comment]
@sxjscience FYI

File: src/operator/tensor/indexing_op.cu [hunk 2]
```diff
@@ -545,6 +545,247 @@ void TakeOpForward<gpu>(const nnvm::NodeAttrs& attrs,
...
+    mean = (lower_bound + upper_bound) / 2;
+    if (id <= sorted_data[mean])
+      upper_bound = mean;
+    else
+      lower_bound = mean + 1;
+  }
+  bool found_row = (sorted_data[lower_bound] == id);
+
+  if (id < vocab_dim) {
+    bounds[id] = (found_row) ? lower_bound : -1;
```
Comment (ptrendx) [inline src/operator/tensor/indexing_op.cu]: Maybe early exit? How long is this kernel?

File: src/operator/tensor/indexing_op.cu [hunk 3]
```diff
@@ -545,6 +545,247 @@ void TakeOpForward<gpu>(const nnvm::NodeAttrs& attrs,
...
+  LType Lvalue[1];
+  DType* Dvalues = reinterpret_cast<DType*>(Lvalue);
+
+  for (index_t row=0; row < rows_per_block; ++row) {
+    IType my_row = blockIdx.x * rows_per_block + row;
+    if (my_row < vocab_dim) {
+      // Read lower and upper bounds for current row
+      IType lower_bound = index_bounds[my_row];
+      IType upper_bound = index_bounds[vocab_dim + my_row];
+      int nOccurrences = (lower_bound != -1) ? (upper_bound - lower_bound + 1) : 0;
```
Comment (ptrendx) [inline src/operator/tensor/indexing_op.cu]: Super small hack and probably not worth it, but if you set `upper_bound` to -2 if you did not find the lower bound, then you would not need this conditional here.

File: src/operator/tensor/indexing_op.cu [hunk 4]
```diff
@@ -545,6 +545,247 @@ void TakeOpForward<gpu>(const nnvm::NodeAttrs& attrs,
...
+
+      for (index_t emb_id=threadIdx.x; emb_id < aligned_emb_dim; emb_id += blockDim.x) {
+        // Initialize grad_in
+        if (req == kAddTo) {
+          grad_in_row[threadIdx.x] = aligned_grad_in[my_row * aligned_emb_dim + emb_id];
+        } else {
+          grad_in_row[threadIdx.x] = 0.0;
+        }
+        // Add all rows from grad_out according to indices in data
+        if (nOccurrences) {
```
Comment (ptrendx) [inline src/operator/tensor/indexing_op.cu]: Does this help? It should be already handled the same way by the for loop.

File: src/operator/tensor/indexing_op.cu [hunk 5]
```diff
@@ -545,6 +545,247 @@ void TakeOpForward<gpu>(const nnvm::NodeAttrs& attrs,
...
+  // Find lower & upper bounds of each possible index
+  const int threads_block_bounds = 128;
+  const int nblocks_bounds = (vocab_dim + threads_block_bounds - 1) / threads_block_bounds;
+  EmbeddingFindBounds<<<nblocks_bounds, threads_block_bounds, 0, Stream<gpu>::GetStream(s)>>>(
+                  sorted_data.dptr_, bounds_index.dptr_, data_dim, vocab_dim);
+
+  // Compute Gradient
+  int ltype = mxnet::common::cuda::get_load_type(embbedding_dim * sizeof(DType));
+  MXNET_LOAD_TYPE_SWITCH(ltype, LType, {
+    const int rows_per_block = 1;
```
Comment (ptrendx) [inline src/operator/tensor/indexing_op.cu]: Do you support `rows_per_block != 1` in the kernel? If not, then it should be removed.

**Comment by sxjscience** [comment]
Nice! LGTM. So the BinarySearch version of FindBounds has complexity O(|V| log |N|) where |V| is the vocabulary size and |N| is the number of indices. I guess our initial version has complexity O(|N|) for finding the boundaries. Thus, in some workloads (in which |N| is small), the O(N) version might be faster.

**Comment by ptrendx** [comment]
Moises did a performance comparison between the new version and both the old one and the old buggy one. The new kernel is faster than the old working version in all cases and ~same in speed as the buggy one. The biggest performance change is seen actually when changing how many different elements are in the input data (as small number of distinct elements limits parallelism in the backward pass).

**Pull request merged**